### DISCUSSION TITLE
NPM Clean Install Issue?

### DISCUSSION THREAD

**Discussion by snarkipus** [eclipse-langium/langium#674] [Q&A] [answered]
Tag: @msujew 

So, I've gotten in the habit of running `npm ci` periodically since I'm prone to making a mess (still haven't sorted out the jest issue).

For some reason, the build failed, and I narrowed it down to an unhandled `undefined` in the workspace manager [link](https://github.com/langium/langium/blob/8c5d43d26a2afd9dadacff02328ff619d74df30b/packages/langium/src/workspace/workspace-manager.ts#L49-L51).

For now, I just did this which passes all of the tests (probably not good), but I'm fairly sure this isn't actually what was intended:
```typescript
services.lsp.LanguageServer.onInitialize(params => {
    if (params.workspaceFolders === undefined) {
        throw new Error('WorkspaceFolder[] is undefined.');
    } else {
        this.folders = params.workspaceFolders;
    }
});
```

Since it's entirely possible I inadvertently broke something, I thought I'd ask first. However, I do see that this was just recently touched in [this commit](https://github.com/langium/langium/commit/14eb53278fa3bf0c6b91ba168993fdf78e3b86f7#diff-9ef9cc03a159c7e21b4350faa99ddd28f97e4f27c934a51508e14531e4f4987b) so maybe it just snuck through and nobody has seen it yet.

**Comment by msujew** [comment, marked as answer]
In what sense did the build break? Did TypeScript complain that `params.workspaceFolders` can be `undefined`? Because that's not the case per the typings. Even if it were during runtime, we deal correctly with it later.

![image](https://user-images.githubusercontent.com/4377073/189551314-c9633468-9a5a-4c8c-8a3e-008b05c8c90d.png)

**Reply by snarkipus** [reply]
For some reason, this is the typings that I get ... maybe it's something weird on my end.

<img width="885" alt="image" src="https://user-images.githubusercontent.com/3688989/189551743-3c5d7222-0e72-412e-8b98-a8d6830795fe.png">

<img width="519" alt="image" src="https://user-images.githubusercontent.com/3688989/189551784-47c773bd-7c7e-4532-bbcd-d37873eacee5.png">

**Reply by msujew** [reply]
Interesting. It seems like Microsoft released a breaking change between `8.0.1` and `8.0.2`, see https://github.com/microsoft/vscode-languageserver-node/pull/948 of the language client/server libraries. Great...

You likely have `8.0.2` locally, while the `main` branch has `8.0.1` locked in. I'll update our dependencies.

**Reply by snarkipus** [reply]
Confirmed via `npm list`.

```term
├─┬ langium-vscode@0.4.0 -> ./packages/langium-vscode
│ ├── @types/vscode@1.71.0 deduped
│ ├── esbuild@0.14.54
│ ├── ignore@5.2.0
│ ├── langium@0.4.0 deduped -> ./packages/langium
│ ├── vscode-languageclient@8.0.2 deduped
│ └── vscode-languageserver@8.0.2 deduped
├─┬ langium@0.4.0 -> ./packages/langium
│ ├── chevrotain@9.1.0 deduped
│ ├── langium-cli@0.4.0 deduped -> ./packages/langium-cli
│ ├── vscode-languageserver-textdocument@1.0.7
│ ├── vscode-languageserver@8.0.2 deduped
│ └── vscode-uri@3.0.3 deduped
```
